spec/system/admin/products_v3/products_spec.rb

```diff
@@ -410,6 +410,179 @@
+        create(:variant,
+               product: product_a,
+               display_name: "Medium box",
+               sku: "APL-01",
+               price: 5.25)
+
+        # to select the default variant
+        within default_variant_selector do
+          page.find(".vertical-ellipsis-menu").click
+          expect(page).to have_css(delete_option_selector)
+        end
+        page.find("div#content").click # to close the vertical actions menu
+
+        within variant_selector do
+          page.find(".vertical-ellipsis-menu").click
+          expect(page).to have_css(delete_option_selector)
+        end
+      end
+    end
+
+    describe "Delete Action" do
+      let!(:variant_a1) {
+        create(:variant,
+               product: product_a,
+               display_name: "Medium box",
+               sku: "APL-01",
+               price: 5.25)
+      }
+      let(:modal_selector) { "div[data-modal-target=modal]" }
+      let(:dismiss_button_selector) { "button[data-action='click->flash#close']" }
+
+      context "when 'keep product/variant' is selected" do
+        it 'should not delete the product/variant' do
+          # Keep Product
+          within product_selector do
+            page.find(".vertical-ellipsis-menu").click
+            page.find(delete_option_selector).click
+          end
+          keep_button_selector = "input[type=button][value='Keep product']"
+          within modal_selector do
+            page.find(keep_button_selector).click
+          end
+
+          expect(page).to_not have_selector(modal_selector)
+          expect(page).to have_selector(product_selector)
+
+          # Keep Variant
+          within variant_selector do
+            page.find(".vertical-ellipsis-menu").click
+            page.find(delete_option_selector).click
+          end
+          keep_button_selector = "input[type=button][value='Keep variant']"
+          within modal_selector do
+            page.find(keep_button_selector).click
+          end
+
+          expect(page).to_not have_selector(modal_selector)
+          expect(page).to have_selector(variant_selector)
+        end
+      end
+
+      context "when 'delete product/variant' is selected" do
+        let(:success_flash_message_selector) { "div.flash.success" }
+        let(:error_flash_message_selector) { "div.flash.error" }
+        it 'should successfully delete the product/variant' do
+          # Delete Variant
+          within variant_selector do
+            page.find(".vertical-ellipsis-menu").click
+            page.find(delete_option_selector).click
+          end
+
+          delete_button_selector = "input[type=button][value='Delete variant']"
+          within modal_selector do
+            page.find(delete_button_selector).click
+          end
+
+          expect(page).to_not have_selector(modal_selector)
+          sleep(0.5) # delay for loading spinner to complete
+          expect(page).to_not have_selector(variant_selector)
```

Comment:
That sleep shouldn't be necessary because Capybara should wait for the variant selector to disappear.
```suggestion
          expect(page).to_not have_selector(variant_selector)
```

Reply:
By the way, I just deleted one of my review comments and noticed that Github uses a simple browser alert instead of a fancy modal. That would have been much easier.

Reply:
> That sleep shouldn't be necessary because Capybara should wait for the variant selector to disappear.

Hmm, I think we are using StimulusReflex and CableReady here for the loading spinner and that works differently than the plain HTML and stuff for Capybara.
If I don't use some delay here then the specs turns out to be flaky i.e. it always passes even when it's not expected. With delay, it passes and  fails accordingly

Reply:
Expectations like `have_selector` still wait some time as well though. Maybe there are other conditions we are not considering. When the modal is open, for example, the variant wouldn't be visible.

The problem with the sleep is that it will probably become flaky as well. Sometimes CI can be very slow and then 0.5 is not enough. We could wait longer but that delays test execution even when the machine is fast. A dilemma.

Reply:
> The problem with the sleep is that it will probably become flaky as well. Sometimes CI can be very slow and then 0.5 is not enough. We could wait longer but that delays test execution even when the machine is fast. A dilemma.

Yes, that's concerning. Maybe I'm missing something here. Let dig into it furthur.

Reply:
I've fixed this here: https://github.com/openfoodfoundation/openfoodnetwork/pull/11846/commits/c5a86341dc2f5400af806a94ff3d913f5ded39d1

I added the sleep because I wanted to make sure that the below shown products loading spinner is hidden before any expectation:
![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/59338032/87434a6e-f849-4fe4-9933-e9842114bafe)
Now I've replaced this with more explicit logic. Please review it and let me know what you think about it.